File: backend/models/emotionClassifier.js
class MusicEmotionClassifier {
  constructor() {
    this.emotions = {
      HAPPY: 'happy',
      SAD: 'sad',
      ENERGETIC: 'energetic',
      CALM: 'calm'
    };
  }

  classifyEmotion(audioFeatures) {
    const { valence, energy, danceability, acousticness } = audioFeatures;

    if (valence > 0.6 && energy > 0.6) {
      return this.emotions.HAPPY;
    }
    
    if (energy > 0.7 && danceability > 0.6) {
      return this.emotions.ENERGETIC;
    }
    
    if (valence < 0.4 && energy < 0.5) {
      return this.emotions.SAD;
    }
    
    if (acousticness > 0.5 && energy < 0.5) {
      return this.emotions.CALM;
    }

    if (valence > 0.5) {
      return energy > 0.5 ? this.emotions.HAPPY : this.emotions.CALM;
    } else {
      return energy > 0.5 ? this.emotions.ENERGETIC : this.emotions.SAD;
    }
  }

  getTargetFeatures(emotion) {
    const targets = {
      [this.emotions.HAPPY]: {
        target_valence: 0.8,
        target_energy: 0.7,
        min_valence: 0.6,
        min_energy: 0.5
      },
      [this.emotions.SAD]: {
        target_valence: 0.2,
        target_energy: 0.3,
        max_valence: 0.4,
        max_energy: 0.5
      },
      [this.emotions.ENERGETIC]: {
        target_energy: 0.9,
        target_danceability: 0.8,
        min_energy: 0.7
      },
      [this.emotions.CALM]: {
        target_energy: 0.3,
        target_acousticness: 0.7,
        max_energy: 0.5
      }
    };

    return targets[emotion] || targets[this.emotions.CALM];
  }

  calculateEmotionScore(audioFeatures, targetEmotion) {
    const targets = this.getTargetFeatures(targetEmotion);
    let score = 0;
    let count = 0;

    if (targets.target_valence !== undefined) {
      score += (1 - Math.abs(audioFeatures.valence - targets.target_valence)) * 100;
      count++;
    }
    if (targets.target_energy !== undefined) {
      score += (1 - Math.abs(audioFeatures.energy - targets.target_energy)) * 100;
      count++;
    }

    return Math.round(score / count);
  }
}

module.exports = MusicEmotionClassifier;

File: backend/evaluation/emotionClassifier.py
class MusicEmotionClassifier:
    """음악 감정 분류기 (Python 버전)"""
    
    def classifyEmotion(self, audioFeatures):
        """오디오 특성을 기반으로 감정 분류"""
        valence = audioFeatures.get('valence', 0)
        energy = audioFeatures.get('energy', 0)
        danceability = audioFeatures.get('danceability', 0)
        acousticness = audioFeatures.get('acousticness', 0)
        
        # Happy: 높은 valence + 높은 energy
        if valence > 0.6 and energy > 0.5:
            return 'happy'
        
        # Sad: 낮은 valence + 낮은 energy
        elif valence < 0.4 and energy < 0.5:
            return 'sad'
        
        # Energetic: 높은 energy + 높은 danceability
        elif energy > 0.7 and danceability > 0.6:
            return 'energetic'
        
        # Calm: 낮은 energy + 높은 acousticness
        elif energy < 0.5 and acousticness > 0.5:
            return 'calm'
        
        # 기본값: valence와 energy로 판단
        elif valence >= 0.5:
            return 'happy'
        else:
            return 'sad'
    
    def calculateEmotionScore(self, audioFeatures, targetEmotion):
        """특정 감정에 대한 매칭 점수 계산 (0-100)"""
        valence = audioFeatures.get('valence', 0)
        energy = audioFeatures.get('energy', 0)
        danceability = audioFeatures.get('danceability', 0)
        acousticness = audioFeatures.get('acousticness', 0)
        
        if targetEmotion == 'happy':
            score = (valence * 0.4 + energy * 0.3 + danceability * 0.2 + (1 - acousticness) * 0.1) * 100
        elif targetEmotion == 'sad':
            score = ((1 - valence) * 0.4 + (1 - energy) * 0.3 + acousticness * 0.2 + (1 - danceability) * 0.1) * 100
        elif targetEmotion == 'energetic':
            score = (energy * 0.4 + danceability * 0.3 + valence * 0.2 + (1 - acousticness) * 0.1) * 100
        elif targetEmotion == 'calm':
            score = ((1 - energy) * 0.4 + acousticness * 0.3 + (1 - danceability) * 0.2 + valence * 0.1) * 100
        else:
            score = 50
        
        return min(100, max(0, score))